server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/NodeReplacementAllocationDecider.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.routing.allocation.decider;
+
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.SingleNodeShutdownMetadata;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.routing.RoutingNode;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.allocation.RoutingAllocation;
+
+import java.util.Map;
+import java.util.Optional;
+
+public class NodeReplacementAllocationDecider extends AllocationDecider {
+
+    public static final String NAME = "node_replacement";
+
+    static final Decision NO_REPLACEMENTS = Decision.single(Decision.Type.YES, NAME,
+        "neither the source nor target node are part of an ongoing node replacement (no replacements)");
+
+    @Override
+    public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        if (replacementOngoing(allocation) == false) {
+            return NO_REPLACEMENTS;
+        } else if (replacementFromSourceToTarget(allocation, shardRouting.currentNodeId(), node.node().getName())) {
+            return Decision.single(Decision.Type.YES, NAME,
+                "node [%s] is replacing node [%s], and may receive shards from it", shardRouting.currentNodeId(), node.nodeId());
+        } else if (isReplacementSource(allocation, shardRouting.currentNodeId())) {
+            return Decision.single(Decision.Type.NO, NAME,
+                "node [%s] is being replaced, and its shards may only be allocated to the replacement target [%s]",
+                shardRouting.currentNodeId(), getReplacementName(allocation, shardRouting.currentNodeId()));
+        } else if (isReplacementSource(allocation, node.nodeId())) {
+            return Decision.single(Decision.Type.NO, NAME,
+                "node [%s] is being replaced by [%s], so no data from other node [%s] may be allocated to it",
```

Comment:
I think users seeing `so no data from other node [null]` (for unassigned shards) may be confused about the null, considering it a bug? I think simplifying it to:

```suggestion
                "node [%s] is being replaced by [%s], so no data may be allocated to it",
```